### PR TITLE
Cleanly shutdown cleanup go-routine

### PR DESCRIPTION
**Pull request by alok87**
Cleanly shutdown go routine. Big cache was made when go did not have ctx support

Fixes https://github.com/allegro/bigcache/issues/336

### CODE REVIEW

File: bigcache.go [hunk 1]
```diff
@@ -41,11 +42,11 @@ const (
 )
 
 // NewBigCache initialize new instance of BigCache
-func NewBigCache(config Config) (*BigCache, error) {
```
Comment (janisz) [inline bigcache.go:44]: I think we should keep the old constructor for backward compatibility but mark it deprecated

File: bigcache_bench_test.go [hunk 2]
```diff
@@ -16,7 +17,7 @@ func BenchmarkWriteToCacheWith1Shard(b *testing.B) {
 
 func BenchmarkWriteToLimitedCacheWithSmallInitSizeAnd1Shard(b *testing.B) {
 	m := blob('a', 1024)
-	cache, _ := NewBigCache(Config{
+	cache, _ := NewBigCache(context.TODO(), Config{
```
Comment (janisz) [inline bigcache_bench_test.go:20]: I think in tests its fine to use `context.Background()`

File: bigcache.go (hunk 1 above)
Comment (alok87) [reply bigcache.go:44]: name suggestion for the new function `NewBigCacheWithContext()`? 

and what should we pass in the old function TODO() or Background() is fine?

Comment (alok87) [reply bigcache.go:44]: or just `New()` as the package name is bigcache

`bigcache.New()`

Comment (janisz) [reply bigcache.go:44]: Let's go with `WithContext` suffix as this is how std lib added context https://godocs.io/net/http#NewRequestWithContext

Comment (alok87) [reply bigcache.go:44]: Both background and todo returns the same empty context
```
var (
	background = new(emptyCtx)
	todo       = new(emptyCtx)
)
```
https://go.dev/src/context/context.go

Comment (cristaloleg) [reply bigcache.go:44]: It's not about implementation, it's about readability 😉

File: bigcache.go [hunk 3]
```diff
@@ -40,12 +41,19 @@ const (
...
 )
 
+// New initialize new instance of BigCache
+func New(ctx context.Context, config Config) (*BigCache, error) {
+	return newBigCache(ctx, config, &systemClock{})
+}
+
 // NewBigCache initialize new instance of BigCache
+// deprecated, please use New(ctx, config) instead, New takes in context and can gracefully
+// shutdown with context cancellations
```
Comment (janisz) [inline bigcache.go:51]:
```suggestion
// NewBigCache initialize new instance of BigCache
//
// Deprecated: NewBigCache is deprecated, please use New(ctx, config) instead, New takes in context and can gracefully
// shutdown with context cancellations
```
https://github.com/golang/go/wiki/Deprecated

**Review by janisz — CHANGES_REQUESTED** [review]

File: bigcache.go (hunk 3 above)
Comment (alok87) [reply bigcache.go]: Fixed

**Review by janisz — APPROVED** [review]

File: bigcache.go (hunk 1 above)
Comment (alok87) [reply bigcache.go:44]: Saw the comment in `WithContext` now, went ahead with `New()`, looks more redable to me. Let me know if it requires a change.

Std lib follows both, see this https://pkg.go.dev/errors#New 
I think it depends on case by case basis. Also for `http.New()` wont really work as it is a multiple object inside http.

**Pull request merged**